### PR TITLE
Node revamp: Introduction to Express

### LINKED ISSUE #27591: New Lesson: Introduction to Express
## Lesson overview

Introduce Express and related concepts that will carry learners through the course.

## Lesson outline

The following items define the scope of the lesson. The lesson outline is flexible; you can improve, expand, and omit items whilst writing lesson content (make sure the pull request description mentions that).

- Make a case for we need Express e.g. writing vanilla Node servers verbose, express popular, express minimalistic
- Go over the request -> response journey in Express, discuss middlewares
- Code example that sets up a simple express project and complements all the concepts above

## Acceptance criteria

If the requirements here are not met, the work effort is not complete.

- [ ] Lesson conforms to the outline above (unless outline has been modified).
- [ ] Lesson follows our [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md).

### Get the lesson template

Download the lesson template using the following command (replace `lesson_name` with the actual lesson name):

```bash
curl -o <lesson_name>.md https://raw.githubusercontent.com/TheOdinProject/curriculum/main/templates/lesson-template.md 
```

### PR DESCRIPTION
**Pull request by MaoShizhong**
## Because
Draft new Express intro lesson as part of the Node revamp project.

The primary purpose of this PR's initial state is not to submit a "final product" but to begin putting bullet points and otherwise previously brainstormed ideas into context in lesson form. That way, it may be easier to begin making tangible progress as the actual course content begins to take shape. This will also allow for any feedback/review/changes to be kept in a single place and make the refinement process smoother.

#### Points to note (as of the initial PR state):
- The assignment is only 1 step and reuses the link for the current course's Basic Informational Site project. If this step remains, the link will of course require updating to whatever the new link is.
- No knowledge checks Qs/additional resources added yet.
- Unsure if Nodemon is most appropriate here. I feel it's appropriate to introduce as early into Express as possible, but not fully convinced.

## This PR
- Provides an `Introduction to Express` lesson

## Issue
Closes #27591

## Pull Request Requirements
-   [ ] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
-   [ ] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [ ] The `Because` section summarizes the reason for this PR
-   [ ] The `This PR` section has a bullet point list describing the changes in this PR
-   [ ] If this PR addresses an open issue, it is linked in the `Issue` section
-   [ ] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [ ] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

**Comment by MaoShizhong** [comment]
@01zulfi I think I would need to explore things a bit more before I reach a stage where it's "ready for review", but while it's in draft I would definitely appreciate any initial feedback. It can help shape the direction it goes in/gives me things to consider that I may not have considered etc.

**Comment by MaoShizhong** [comment]
@01zulfi I think at this point, it's ready for an initial review.

Important notes/thoughts:
- My aim for this lesson is very literally to just introduce what Express is and setting up the most basic Express app in order to demonstrate the request->response journey.
- Any details of routes/MVC are to be kept for the appropriate dedicated lesson(s). This lesson should only include what is necessary to demonstrate the above point, and allow for the basic information site project rewrite.
- Because of the limited scope of this lesson, I don't think there needs to be much more in the assignment section. The two articles from the current course go through a lot of redundant stuff (Node/npm etc.) and/or stuff better placed in a later lesson. The key info from those have now been written in-house.
- No additional resources set, mainly for similar reasons to the above. The current course lesson's additional resources go into way more detail than is necessary for this lesson, and many of those details are already (and should be) covered in future lessons.
- The link for the basic information project in the assignments section will of course need changing to the new version when that time comes. The current old course link is a placeholder.

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:3]: Do you mean previous lesson or previous section?

Also, not sure about the word "procedures", let's go with "use cases" or something similar: 
```suggestion
In the previous lesson, you got up and running with Node. You learned how to read and write data from files on the server, and use those to serve up a multi-page website. You may have found this a somewhat verbose and cumbersome process and may have even wondered how on Earth we would manage more complex use cases.
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:5]: Wording suggestion: 
```suggestion
In this course, we are going to be using a backend framework called [Express](https://expressjs.com/). It exposes an easy to use API that handles a lot of the implementation details for us. Express itself is an intentionally barebones and unopinionated framework; it allows us to do many things how we want, and to extend it with only the features we need. While this gives us great flexibility in how we do things, it can be a little tricky deciding between multiple viable options at times.
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:22]: We could just ask learners to run `npm init -y`. Your call.

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:40]: Feels to verbose, and not sure if we need to add the `require('express')();` thing. Wording suggestion:
```suggestion
Let's break this down. We import and call `express` to initialize the `app` variable.  This is our server.
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:44]: Wording suggestion: 
```suggestion
Finally, we tell our server to listen for requests on port 3000 of our [localhost](https://simple.wikipedia.org/wiki/Localhost) (which is basically just the computer's local connection). 3000 is an arbitrary and somewhat conventional default choice, but you can technically use any unused port (e.g. Vite's dev server uses port 5173 by default). Back in your terminal, if you run `node app.js` then all being well, you should see `My first Express app - listening on port 3000!` logged. Congratulations! Your first Express server is now running.
```

Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:44]: What about adding in a lesson note titled "The Port Variable". It says something like "`port` is usually a environment variable and we'll cover environment variables later. For now, it's perfectly fine to have it as a global variable."

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:52]: Wording suggestion:
```suggestion
Once you navigate to `http://localhost:3000/`, you should see `Hello, world!` appear in the window. Magic, right?
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:54]: Wording suggestion:
```suggestion
When our server receives our `GET` request, Express stores the request in a [`request` object](https://expressjs.com/en/4x/api.html#req). This request gets passed through a chain of functions we call `middleware functions` until eventually, a middleware function tells Express to respond to the request.
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:56]: We could be more explicit here: 
```suggestion
In our example, the request comes through as a `GET` request to the `/` route. This matches the route handler we have in our `app.js` file.
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:62]: Wording suggestion: 
```suggestion
We will discuss routes and route handlers in more detail in a later lesson, but to summarise the above line, it tells Express "if a `GET` request comes through to the `/` route, run the request through the following chain of middleware functions". Here, we only have a single function.
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:74]: Love this suggestion. Though I wonder if Node's built-in watch will suffice for the course's use case?

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:80]: We can add some reading resources from the Express docs. This will be super valuable for our learners.

This are the sections I have in mind (as they are already covered in the lesson):

- https://expressjs.com/en/4x/api.html#app.get.method
- https://expressjs.com/en/4x/api.html#app.listen
- https://expressjs.com/en/4x/api.html#res.send

**Review by 01zulfi — CHANGES_REQUESTED** [review]
@MaoShizhong great work here 🔥 🔥 

Left a few comments below. The comments marked "Wording suggestion:" are subjective and up to you if you want to use my suggestion or modify it on your own.

File: nodeJS/express/introduction_to_express.md
Comment (Asartea) [inline nodeJS/express/introduction_to_express.md:54]: This is one of several places with a link where part of the link text is code formatted; this [tends to look somewhat wonky](https://www.theodinproject.com/lessons/preview?uuid=bd4a27c8-f851-4a97-ae83-1990be86e4a9) because the underline will get split between the code and the rest of the text, making it look like they are separate links

File: nodeJS/express/introduction_to_express.md
Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md:5]: Like the more streamlined wording. I've gone for even more streamlined as I'm not sure about explicitly saying "this is easy to use/has an easy to use XYZ", but the latter part is definitely more than sufficient.

File: nodeJS/express/introduction_to_express.md
Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md:22]: We could. I also just noticed that `package.json` defaults to `"main": "index.js"`.
I don't think this should be any issue really since AFAIK it's only for creating an npm package so it knows what entry point to use when the package is imported.

I guess fine to leave `main` at that and just don't mention it?

`npm init -y` would probably be easier.

File: nodeJS/express/introduction_to_express.md
Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md:52]: Oooh, I *like* it. Very snazzy.

File: nodeJS/express/introduction_to_express.md
Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md:74]: I thought about that but given it's still experimental in LTS (heck, even in v21), I opted against it as I feel the curriculum should stick to recommending stable features. Similarly to how Jest's native ESM support is still experimental due to the relevant Node APIs being unstable - hence it'd be better to stick with the very quick babel setup for Jest/ESM usage.

File: nodeJS/express/introduction_to_express.md
Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md:44]: A lesson note sounds good. Might help throw in some visual variance as well.

File: nodeJS/express/introduction_to_express.md
Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md:80]: I was of the mind that at least for app.get and res.send, they'd be more suitable in the routes lesson perhaps? Since that's when we'd ideally be breaking down the anatomy of a route handler and middleware function, where both of these can be discussed in more detail.

And I feel app.listen as a sole "extra reading" feels a bit off, though that may just be me.

File: nodeJS/express/introduction_to_express.md
Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md:40]: In fairness, not the greatest show but I actually completely forgot I wrote this extra `require()()` stuff in. The initial idea was to answer "bu why do we need an intermediary variable?" but in hindsight, don't think that'd be a common question and anyone who does ask that in the community can just be asnwered that/told to keep going.

File: nodeJS/express/introduction_to_express.md
Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md]: Fair shout! Looking at it again, it would perfectly suffice without the backticks anyway given they're actually the only places I had those words backticked. The link visuals are sufficient.

File: nodeJS/express/introduction_to_express.md
Comment (cakegod) [inline nodeJS/express/introduction_to_express.md:48]:
```suggestion
Now that our server is up and running on port 3000, it's time to send it a request! Open your browser and navigate to `http://localhost:3000/.` Don't worry if you forget the trailing slash `/` at the end – the browser will silently add it for you. This action instructs your browser to send a `GET` request to the `/` route of whatever server is running at port 3000 on our localhost (which happens to be our Express server!), and display in the window whatever it receives in the response.
```

**Review by cakegod — COMMENTED** [review]
I've leaving you some suggestions for half the lesson.

**Comment by MaoShizhong** [comment]
@cakegod Thank you for the suggestions - agree with many of them as general improvements.
Some I personally felt were more alternative wordings than improvements so did not implement those ones. Also I specifically linked the simple english wiki for localhost because I like its definition, and there was no other stuff to potentially rabbit hole into unlike the normal wiki page for it.

**Comment by cakegod** [comment]
@MaoShizhong That's fine, they're merely suggestions. I'll check the rest of the lesson tomorrow.

File: nodeJS/express/introduction_to_express.md
Comment (cakegod) [inline nodeJS/express/introduction_to_express.md:66]:
```suggestion
Express takes the callback function we gave it and passes a request object into the first parameter (commonly named `req` by convention), and a [response object](https://expressjs.com/en/4x/api.html#res) into the second parameter (commonly named `res`). Our callback tells the response object to respond to the request by `.send`ing the string `"Hello, world!"`.
```

File: nodeJS/express/introduction_to_express.md
Comment (cakegod) [inline nodeJS/express/introduction_to_express.md:68]:
```suggestion
Since there is no more code to run, the function returns and Express responds to the request, effectively ending the request-response cycle. Meanwhile, the browser receives our server's response, which is our "Hello, world!" string, and displays it on the screen. We could send nearly anything in a response. We could even [tell Express to send a file](https://expressjs.com/en/api.html#res.sendFile).
```

File: nodeJS/express/introduction_to_express.md
Comment (cakegod) [inline nodeJS/express/introduction_to_express.md:72]:
```suggestion
When you run your server with `node app.js`, any changes to JavaScript and JSON files in your project directory won't be reflected automatically unless you manually interrupt and rerun `node app.js`. To avoid this manual process, you can use [Nodemon](https://www.npmjs.com/package/nodemon) to run your "app.js" file. Nodemon automatically restarts and rebuilds your site when you make changes to files, similar to Webpack and Vite's dev servers.
```

**Review by cakegod — COMMENTED** [review]
More suggestions.

File: nodeJS/express/introduction_to_express.md
Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md:64]: Definitely agree with the improved clarity.

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [reply nodeJS/express/introduction_to_express.md]: I've been thinking about this. I'm inclined towards covering environment variables earlier, since learners will be dealing with them during deployments.

I think we could even cover env vars before the Express section altogether.

There's also a relevant issue on it https://github.com/TheOdinProject/curriculum/issues/26851

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [reply nodeJS/express/introduction_to_express.md]: Makes sense

File: nodeJS/express/introduction_to_express.md
Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md]: Environment variables being covered before Express does make a heap load of sense, and I'm very on board with that idea.
One of the problems with the current order is that the Local Library tutorial doesn't expect you to push to a repo until the end, so it instructs you to hardcode a DB connection string. Only when it reaches the end with deployment does it tell you to swap it for an env variable, push to GitHub, then set the env value in the host. But then people will often push to a repo only to discover later they probably weren't supposed to, and now have their connection string publicly visible. So .env mention is being brought forwards as a bandaid, but then that makes the security lesson kinda moot since most of it is about .env ...

For now, how would you prefer me to handle env variables in this lesson? Make the wording change with the assumption that env has already been covered pre-Express? Then it'll be resolved when we do get to introducing it in the pre-Express section in a later milestone.

Comment (01zulfi) [reply nodeJS/express/introduction_to_express.md]: Yeah, we can assume learners know about env vars in this lesson.

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [reply nodeJS/express/introduction_to_express.md:80]: Rather than specific links, we can just link the Express site https://expressjs.com/ and ask learners to explore the documentation on their own with wording similar to "Spend a few minutes exploring the Express documentation so you can get a feel for it. We'll be referencing a lot of content from the docs in the next lessons"

Comment (MaoShizhong) [reply nodeJS/express/introduction_to_express.md:80]: Sensible - I'll add that as an assignment step before the basic info site rewrite.

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:7]: I'm not sure if this sentence belongs here, makes the lesson sound like an intro to a course. Don't feel strongly towards it, your call.

```suggestion
Going forward, we will be diving into how we can use Express in various ways, such as to create a full-stack application using the Model View Controller (MVC) pattern, as well as to create a REST API just like ones you will have used before for things such as the Weather App or the React Shopping Cart projects. There is a lot to take in, so take it steady, and do not be afraid to ask our community for help in [our Discord server](https://discord.gg/theodinproject)!
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:18]: The first two items don't add a lot of value, they could be removed altogether.
```suggestion
- Set up a basic Express server.
- Describe how an incoming request is processed and ultimately responded to by the server.
- Describe what a middleware function is.
- Describe how to use `nodemon` to automatically restart your application when a change is made.
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:44]: Personal preference: We make the congratulations sentence its own para, your call though
```suggestion
Finally, we tell our server to listen for incoming requests requests on port 3000 of our [localhost](https://simple.wikipedia.org/wiki/Localhost) (which is basically just the computer's local connection). While port 3000 is the default choice, you can use any unused port (for example, Vite's dev server uses port 5173 by default). Back in your terminal, if you run `node app.js` then all being well, you should see `My first Express app - listening on port 3000!` logged.

Congratulations! Your first Express server is now running.
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:66]: Wording suggestion
```suggestion
Express takes the callback function we gave it and passes the request object into the first parameter (conventionally named `req`), and a [response object](https://expressjs.com/en/4x/api.html#res) into the second parameter (`res`). Our callback tells the response object to respond to the request by `.send`ing the string `"Hello, world!"`.
```

File: nodeJS/express/introduction_to_express.md
Comment (01zulfi) [inline nodeJS/express/introduction_to_express.md:89]: Not sure if "What is localhost?" is an important learning of the lesson
```suggestion
```

**Review by 01zulfi — CHANGES_REQUESTED** [review]
Minor nits

**Comment by MaoShizhong** [comment]
Sensible suggestions @01zulfi :+1:

**Comment by MaoShizhong** [comment]
Changes applied - definitely helps with streamlining things.

The port/env thing, I applied as a lesson note. For simplicity's sake, I didn't want to include env and dotenv stuff in the example itself, because then it kinda opened a can of worms as to "well now the instructions need to say to make a `.env` file and to show what's in it. Oh look, now I should probably say to `.gitignore` the `.env` file as well" etc.

Thought the lesson note mentioning the fixed value is for demonstration simplicity, and explaining the reasons why we might prefer an env variable for the port value should suffice.

Also linked to the Express 4.X docs specifically, as that's the version npm will install (5.X still in beta). Linking the main website will potentially have a handful of people go looking at the 5.X docs where some things might differ.

**Review by 01zulfi — APPROVED** [review]
@MaoShizhong lgtm 🚀 can't wait for the lesson to go live

let's hold off on merging the PR for now

**Pull request merged**
